### DISCUSSION TITLE
Implement mongose as a MicroPython library

### DISCUSSION THREAD

**Discussion by AmirHmZz** [cesanta/mongoose#1671] [Feature requests]
I'm using [MicroPython](https://github.com/micropython/micropython) for about one year. There's lack of a complete web solution in micropython and I think mongose can easily fit in that situation. I've also opened a feature request on https://github.com/micropython/micropython/issues/8737.

**Comment by cpq** [comment]
Micropython has basic networking support. It has a simple HTTP client / server example in the docs.
Can you elaborate, what extra functionality you're missing where Mongoose can be beneficial?

**Comment by AmirHmZz** [comment]
Thanks for your reply. 

- As you mentioned `Micropython` has only basic low-level [`socket`](https://docs.micropython.org/en/latest/library/socket.html) functionality, So developers will suffer from developing their own `HTTP` server / client. In fact Micropython doesn't have a standard `HTTP` functionality anyway.
- Micropython doesn't support `websocket` ( It has a built-in support which is used for its own [`WebREPL`](https://docs.micropython.org/en/latest/esp8266/tutorial/repl.html#webrepl-a-prompt-over-wifi) but APIs are not exposed to be used )
- Micropython doesn't support MQTT at all.
- It's socket layer DNS resolver [is not asynchronos](https://docs.micropython.org/en/latest/library/uasyncio.html#uasyncio.open_connection).

There's a point : There are libraries for everything (except last one) that I mentioned above but :

- They are not officially maintained by Micropython
- They are non-standard
- All of them are written in `Python` so they don't have small RAM footprints and also they are not as performant as Mongoose.

**Reply by cpq** [reply]
@AmirHmZz thanks. If mongoose had a micropython port/binding, would micropython project integrate it?
Considering that Mongoose is dual licensed, GPLv2/commercial.

**Reply by AmirHmZz** [reply]
I'm not the one who can answer this. I've opened a feature request issue in Micropython repo https://github.com/micropython/micropython/issues/8737. If everything seems ok to you please, come to this issue and negotiate about integration and license stuff.

**Comment by AmirHmZz** [comment]
@cpq Thanks for participating :) So according to https://github.com/micropython/micropython/issues/8737#issuecomment-1214501358 what will be your final descision?
I should admit that Micropython community will be appreciative and glad If this happens.

**Comment by cpq** [comment]
The micropython integration looks interesting indeed.

It needs a bit of work to get done though; so unless someone is willing to invest an effort, it is unlikely to happen soon.
However, if we had a commercial customer asking for micropython support, that'd be a strong motivation for quick implementation.

**Comment by scaprile** [comment]
#2450

**Reply by AmirHmZz** [reply]
@scaprile It's 2025 and still micropython doesn't have a proper http toolkit solution yet...

**Reply by cpq** [reply]
@AmirHmZz isn't https://github.com/cesanta/mongoose/tree/master/tutorials/micropython good enough?

**Reply by scaprile** [reply]
Or (pardon me for being self-referential here) : https://github.com/cesanta/mongoose/blob/e07a8283ec45e6ababb236bd645b8c6c4fecc5cc/tutorials/micropython/README.md?plain=1#L7
https://www.embeddedrelated.com/showarticle/1649.php
https://www.embeddedrelated.com/showarticle/1670.php

Though I think I got your comment as a shame on MicroPython itself.
